src/pubsub.rs

```diff
@@ -0,0 +1,254 @@
+use std::collections::{HashMap, HashSet};
+
+use abscissa_core::{
+    tracing::log::{debug, error},
+    Application,
+};
+use somm_proto::{
+    cosmos_sdk_proto::cosmos::base::abci::v1beta1::TxResponse,
+    pubsub::{
+        query_client::QueryClient, QueryDefaultSubscriptionsRequest,
+        QueryDefaultSubscriptionsResponse, QueryPublisherRequest, QueryPublisherResponse,
+        QuerySubscriberIntentsBySubscriberAddressRequest,
+        QuerySubscriberIntentsBySubscriberAddressResponse, Subscriber,
+    },
+};
+use url::Url;
+use x509_parser::prelude::parse_x509_pem;
+
+use crate::{
+    config::get_delegate_address,
+    error::{Error, ErrorKind},
+    prelude::APP,
+    somm_send,
+};
+
+use self::cache::PublisherTrustData;
+
+pub(crate) mod cache;
+
+/// Retrieves the PEM encoded CA certs for all default subscription publishers.
+pub(crate) async fn get_trust_state() -> Result<Vec<PublisherTrustData<'static>>, Error> {
+    debug!("collecting trust roots from default subscriptions and subscriber intents.");
+    let config = APP.config();
+
+    // load subscription mappings
+    let mut subscription_mappings: HashMap<String, HashSet<String>> = HashMap::default();
+    get_default_subscriptions()
+        .await?
+        .default_subscriptions
+        .iter()
+        .for_each(|s| {
+            if config
+                .pubsub
+                .publisher_domain_block_list
+                .contains(&s.publisher_domain)
+            {
+                debug!(
+                    "publisher domain {} is in the block list. skipping.",
+                    s.publisher_domain
+                );
+
+                return;
```

Comment:
Wouldn't we want the blocklist to be enforced in the authorization handler rather than in the cache? That would have immediate effect, although I suppose a config change requires the process to restart anyway, so I'm not sure there.

In either case, we don't need to process the block list twice, we can create the overall mappings and then filter out blocked entries by deleting the publisher key.

Also one more question: in a for_each like this, is return what we want? I guess it's because it's a code block in the iterator rather than an actual loop?

Reply:
Ah I'll move it outside. 

Yeah because it's a closure, return acts like `continue` inside of the for_each method